From d3aa66d17690513e637ebd58ccb066d0d84f44fd Mon Sep 17 00:00:00 2001
From: Randolph <randolph@andestech.com>
Date: Mon, 14 Oct 2024 10:51:38 +0800
Subject: [PATCH 27/53] swdtb: add generic DMA properties for legacy built-in
 DTBs (#130)

The legacy DMA of some built-in DTBs missed the following properties:

In the MMC node:
dmas = <0x07 0x09>;
dma-names = "rxtx";

In the DMA node:
"#dma-cells = <0x01>;"

Because the legacy FPGA bitmaps lack DMA coherence in the FDT's root,
reading the DMA-coherent type from the MMC node is necessary.

Bugzilla: http://e-andes.andestech.com/bugzilla5/show_bug.cgi?id=31805

Signed-off-by: Randolph <randolph@andestech.com>
Reviewed-on: https://gitea.andestech.com/RD-SW/u-boot.official/pulls/130
Reviewed-by: Leo Yu-Chi Liang <ycliang@andestech.com>
Reviewed-by: Tim Shih-Ting OuYang <tim609@andestech.com>
Co-authored-by: Randolph <randolph@andestech.com>
Co-committed-by: Randolph <randolph@andestech.com>
---
 .../common/swdtb_merger/swdtb_merger.c        |   6 +
 .../common/swdtb_merger/swdtb_merger_fixup.c  | 134 ++++++++++++++++++
 .../common/swdtb_merger/swdtb_merger_fixup.h  |   4 +
 .../common/swdtb_merger/swdtb_tool.c          |  75 +++++++++-
 .../common/swdtb_merger/swdtb_tool.h          |   5 +-
 5 files changed, 221 insertions(+), 3 deletions(-)

diff --git a/board/andestech/common/swdtb_merger/swdtb_merger.c b/board/andestech/common/swdtb_merger/swdtb_merger.c
index d4bbdcd24d80..cedbc7384681 100644
--- a/board/andestech/common/swdtb_merger/swdtb_merger.c
+++ b/board/andestech/common/swdtb_merger/swdtb_merger.c
@@ -53,6 +53,12 @@ struct swdtb_table dtb_merge_table[] = {
 	 "dma type to lcd",		dma_type_to_lcd_fixup},
 	{SWDTB_ENABLE,	1,
 	 "add hpm event",		andespmu_event_fixup},
+	{SWDTB_ENABLE,	1,
+	 "add dma rxtx",		mmc_dma_rxtx_fixup},
+	{SWDTB_ENABLE,	1,
+	 "add dmas",			mmc_dma_dmas_fixup},
+	{SWDTB_ENABLE,	1,
+	 "add dma cell",		dma_cells_fixup},
 };
 
 /*
diff --git a/board/andestech/common/swdtb_merger/swdtb_merger_fixup.c b/board/andestech/common/swdtb_merger/swdtb_merger_fixup.c
index 42913c9d92b7..edea7c7619b5 100644
--- a/board/andestech/common/swdtb_merger/swdtb_merger_fixup.c
+++ b/board/andestech/common/swdtb_merger/swdtb_merger_fixup.c
@@ -792,3 +792,137 @@ int cpu_intc_fixup(void *fdt, int node, char *name, char *value)
 	return compatible_change(fdt, node, name, value, 
 				 old_str, new_str, type);
 }
+
+/**
+ * This fixup will add the dma_names into mmc node.
+ * @return 0 if ok, or FDT_ERR... if not.
+ */
+int mmc_dma_rxtx_fixup(void *fdt, int node, char *name, char *value)
+{
+	const char *target_node = "mmc";
+	const char *dma_names = "dma-names";
+	const char *dma_rxtx = "rxtx";
+	const char *full_name;
+	int ret;
+
+	full_name = fdt_get_name(fdt, node, NULL);
+	if (strncmp(full_name, target_node, strlen(target_node)) == 0) {
+		const u32 *value;
+
+		value = fdt_getprop(fdt, node, dma_names, NULL);
+		if (!value) {
+			show_rawdtb(dma_rxtx, target_node);
+			/* add default property of fixed link */
+			ret = fdt_appendprop_string(fdt, node,
+						    dma_names, dma_rxtx);
+			if (ret) {
+				printf("fdt_appendprop_string: err=%s\n",
+				fdt_strerror(ret));
+				return ret;
+			}
+
+			show_result(fdt, node, dma_names, RESULT_STRING);
+		}
+
+		return ret;
+	}
+
+	return !(RETURN_DONE);
+}
+
+/**
+ * This fixup will add the dmas into mmc node.
+ * dmas has two variable, the first is point to the dma node,
+ * the second is the fixed handshake id.
+ * @return 0 if ok, or FDT_ERR... if not.
+ */
+int mmc_dma_dmas_fixup(void *fdt, int node, char *name, char *value)
+{
+#define DMA_DMAS_HANDSHAKE_ID	0x9
+	const char *target_node = "mmc";
+	const char *dmas = "dmas";
+	const char *full_name;
+	const char *dma_compat = "andestech,atcdmac300";
+	const char *dma_compat_g = "andestech,atcdmac300g";
+
+	int dmanode;
+	int dma_phandle;
+	int ret;
+	fdt32_t dmas_array[2];
+
+	full_name = fdt_get_name(fdt, node, NULL);
+	if (strncmp(full_name, target_node, strlen(target_node)) == 0) {
+		const u32 *value;
+
+		value = fdt_getprop(fdt, node, dmas, NULL);
+		if (value)
+			return RETURN_DONE;
+
+		dmanode = find_node_for_gnog(fdt, dma_compat, dma_compat_g);
+		if (dmanode <= 0)
+			return dmanode;
+
+		dma_phandle = fdt_get_phandle(fdt, dmanode);
+		if (!dma_phandle) {
+			dma_phandle = fdt_create_phandle(fdt, dmanode);
+			if (!dma_phandle) {
+				dtb_error("Error on fdt_get_phandle: %s\n",
+					  fdt_strerror(dma_phandle));
+				return dma_phandle;
+			}
+		}
+
+		dmas_array[0] = cpu_to_fdt32(dma_phandle);
+		dmas_array[1] = cpu_to_fdt32(DMA_DMAS_HANDSHAKE_ID);
+
+		ret = fdt_appendprop(fdt, node, dmas,
+				     dmas_array, 2 * sizeof(fdt32_t));
+		if (ret) {
+			printf("fdt_appendprop: err=%s\n", fdt_strerror(ret));
+			return ret;
+		}
+
+		show_result_multi_num(fdt, node, dmas, 2);
+
+		return ret;
+	}
+
+	return !(RETURN_DONE);
+}
+
+/**
+ * This fixup will add the #dma-cells into dma node.
+ * The dma-cells are for the format of the 'interrupts' property.
+ * @return 0 if ok, or FDT_ERR... if not.
+ */
+int dma_cells_fixup(void *fdt, int node, char *name, char *value)
+{
+	const char *target_node = "dma";
+	const char *dma_cells = "#dma-cells";
+	const u32 cells_value = 1;
+	const char *full_name;
+	int ret;
+
+	full_name = fdt_get_name(fdt, node, NULL);
+	if (strncmp(full_name, target_node, strlen(target_node)) == 0) {
+		const u32 *value;
+
+		value = fdt_getprop(fdt, node, dma_cells, NULL);
+		if (!value) {
+			show_rawdtb(dma_cells, target_node);
+			ret = fdt_appendprop_cell(fdt, node,
+						  dma_cells, cells_value);
+			if (ret) {
+				printf("fdt_appendprop_string: err=%s\n",
+				fdt_strerror(ret));
+				return ret;
+			}
+
+			show_result(fdt, node, dma_cells, RESULT_NUMERAL);
+
+			return ret;
+		}
+	}
+
+	return !(RETURN_DONE);	
+}
diff --git a/board/andestech/common/swdtb_merger/swdtb_merger_fixup.h b/board/andestech/common/swdtb_merger/swdtb_merger_fixup.h
index e4acf2f3c727..7966a486bc89 100644
--- a/board/andestech/common/swdtb_merger/swdtb_merger_fixup.h
+++ b/board/andestech/common/swdtb_merger/swdtb_merger_fixup.h
@@ -54,4 +54,8 @@ int insert_inject_node(void *fdt, int node, char *name, char *value);
 int dma_global_pool_fixup(void *fdt, int node, char *name, char *value);
 
 int andespmu_event_fixup(void *fdt, int node, char *name, char *value);
+
+int mmc_dma_rxtx_fixup(void *fdt, int node, char *name, char *value);
+int mmc_dma_dmas_fixup(void *fdt, int node, char *name, char *value);
+int dma_cells_fixup(void *fdt, int node, char *name, char *value);
 #endif
diff --git a/board/andestech/common/swdtb_merger/swdtb_tool.c b/board/andestech/common/swdtb_merger/swdtb_tool.c
index f07b44d19016..59bda2e6af93 100644
--- a/board/andestech/common/swdtb_merger/swdtb_tool.c
+++ b/board/andestech/common/swdtb_merger/swdtb_tool.c
@@ -113,8 +113,13 @@ int dma_type_to_node(void *fdt, int node, char *name, const char *target_node)
 			global_dma_type = DMA_TYPE_COHERENT;
 			printf("Bitmap's DMA type is %s\n", target_1);
 		} else {
-			global_dma_type = DMA_TYPE_NONCOHERENT;
-			printf("Bitmap's DMA type is %s\n", target_2);
+			/* The AX45MP_4C Gen1 is using the legecy DMA */
+			dma_type_from_mmc_node(fdt);
+
+			if (global_dma_type == DMA_TYPE_INIT) {
+				global_dma_type = DMA_TYPE_COHERENT;
+				printf("Bitmap's DMA type is %s\n", target_2);
+			}
 		}
 	}
 
@@ -142,6 +147,53 @@ int dma_type_to_node(void *fdt, int node, char *name, const char *target_node)
 	return ret;
 }
 
+int find_node_for_gnog(void *fdt, const char *node_nog, const char *node_g)
+{
+	int offset_nog;
+	int offset_g;
+	int final_node;
+
+	offset_nog = fdt_node_offset_by_compatible(fdt, 0, node_nog);
+
+	offset_g = fdt_node_offset_by_compatible(fdt, 0, node_g);
+
+	if (offset_nog > 0) {
+		dtb_debug("offset_nog is %d\n", offset_nog);
+		final_node = offset_nog;
+	}
+
+	if (offset_g > 0) {
+		dtb_debug("offset_g is %d\n", offset_g);
+		final_node = offset_g;
+	}
+
+	return final_node;
+}
+
+/*
+ * The dma_type_from_mmc_node function attempts to parse
+ * the dma-coherent property from the MMC node.
+ * In most bitmaps, dma-coherent is located in the root of the FDT.
+ * However, in the A[x]45MP4C_Gen1 bitmap, it is placed in
+ * subnodes such as mmc, mac, and lcd
+ */
+void dma_type_from_mmc_node(void *fdt)
+{
+	const char *mmc_compat_g = "andestech,atfsdc010g";
+	const char *mmc_compat = "andestech,atfsdc010";
+	const char *target_1 = "dma-coherent";
+	int mmc_node;
+	const u32 *target_1_idx;
+
+	mmc_node = find_node_for_gnog(fdt, mmc_compat, mmc_compat_g);
+
+	target_1_idx = fdt_getprop(fdt, mmc_node, target_1, NULL);
+	if (target_1_idx) {
+		global_dma_type = DMA_TYPE_COHERENT;
+		printf("Bitmap's DMA type is %s\n", target_1);
+	}
+}
+
 void show_result(const void *fdt, int node, const char *prop_name, int type)
 {
 #ifdef ANDES_SWDTB_MERGER_CHECK
@@ -179,6 +231,25 @@ void show_result(const void *fdt, int node, const char *prop_name, int type)
 #endif
 }
 
+void show_result_multi_num(const void *fdt, int node,
+			   const char *prop_name, int num_count)
+{
+#ifdef ANDES_SWDTB_MERGER_CHECK
+	const char *data;
+	int idx;
+
+	data = (char *)fdt_getprop(fdt, node, prop_name, NULL);
+	if (!data)
+		dtb_error("Error on fdt_getprop\n");
+
+	printf("Result: %s: %s", prop_name, KYEL);
+	for(idx = 0; idx < num_count; idx++) {
+		printf("%d ", fdt32_to_cpu(*((u32 *)data + idx)));
+	}
+	printf("%s\n", KNRM);
+#endif
+}
+
 int show_reserved_memory_node(const void *fdt, int node,
 			      const char *compatible)
 {
diff --git a/board/andestech/common/swdtb_merger/swdtb_tool.h b/board/andestech/common/swdtb_merger/swdtb_tool.h
index 266a2dffe4ef..1ce192d5274d 100644
--- a/board/andestech/common/swdtb_merger/swdtb_tool.h
+++ b/board/andestech/common/swdtb_merger/swdtb_tool.h
@@ -75,6 +75,8 @@ int compatible_change(void *fdt, int node, char *name, char *value,
 int dma_type_to_node(void *fdt, int node, char *name, const char *target_node);
 int dma_type_fixup(void *fdt, int node, char *name,
 		   char *value, const char *target_node);
+void dma_type_from_mmc_node(void *fdt);
+int find_node_for_gnog(void *fdt, const char *node_nog, const char *node_g);
 
 void show_rawdtb(const char *prop_name, const char *node_name);
 
@@ -82,8 +84,9 @@ void show_rawdtb(const char *prop_name, const char *node_name);
 #define RESULT_STRING	0
 #define RESULT_NUMERAL	1
 void show_result(const void *fdt, int node, const char *prop_name, int type);
+void show_result_multi_num(const void *fdt, int node,
+			   const char *prop_name, int num_count);
 
 int show_reserved_memory_node(const void *fdt, int node,
 			      const char *compatible);
-
 #endif
-- 
2.34.1

